cmd/agent/Dockerfile

```diff
@@ -1,13 +1,18 @@
-FROM alpine:latest as certs
-RUN apk add --update --no-cache ca-certificates
+ARG base_image
+ARG debug_image
 
-FROM scratch
+FROM $base_image AS release
 ARG TARGETARCH=amd64
 ARG USER_UID=10001
-COPY --from=certs /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/ca-certificates.crt
-
-EXPOSE 5775/udp 6831/udp 6832/udp 5778
 COPY agent-linux-$TARGETARCH /go/bin/agent-linux
+EXPOSE 5775/udp 6831/udp 6832/udp 5778/tcp
 ENTRYPOINT ["/go/bin/agent-linux"]
+USER ${USER_UID}
 
+FROM $debug_image AS debug
+ARG TARGETARCH=amd64
+ARG USER_UID=10001
+COPY agent-debug-linux-$TARGETARCH /go/bin/agent-linux
+EXPOSE 5775/udp 6831/udp 6832/udp 5778/tcp 12345/tcp
+CMD ["/go/bin/dlv", "exec", "/go/bin/agent-linux", "--headless", "--listen=:12345", "--api-version=2", "--log"]
```

Comment:
Shouldn't this be an `ENTRYPOINT` instead? Also, you might want to add `--` as the final argument, so that arguments passed when executing the image are relayed to the underlying binary instead of to dlv:

```
ENTRYPOINT ["/go/bin/dlv", "exec", "/go/bin/collector-linux", "--headless", "--listen=:12345", "--api-version=2", "--log", "--"]
```

Like:

```console
$ docker run -p 14250:14250 -p 12345:12345 -e SPAN_STORAGE_TYPE=memory jaegertracing/jaeger-collector:latest-debug --log-level=debug
API server listening at: [::]:12345
2020-10-13T08:25:42Z info layer=debugger launching process with args: [/go/bin/collector-linux --log-level=debug]
```

I was able to connect to the process in the container, but I wasn't able to set a breakpoint via Visual Studio Code. Were you able to accomplish that? Here's the configuration I'm using:

```json
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Launch remote",
            "type": "go",
            "request": "attach",
            "mode": "remote",
            "remotePath": "/go/bin/collector-linux",
            "port": 12345,
            "host": "127.0.0.1",
            "cwd": "/home/jpkroehling/Projects/src/github.com/jaegertracing/jaeger",
        }
    ]
}
```

Reply:
Thank you for the feedback @jpkrohling 

I am not sure but I think since we don't have the source code inside the container that's why we aren't able to set the breakpoint. Let me look into it more.

Reply:
Alright, let me know once you figure it out.

Reply:
I followed this blog: https://golangforall.com/en/post/go-docker-delve-remote-debug.html to debug the breakpoint issue. I think the issue is due to VSCode's go module detection which isn't working properly. When I tried adding breakpoint in VSCode I got this error:

```
2020-10-13T13:17:04Z debug layer=rpc -> *rpc2.CreateBreakpointOut{"Breakpoint":{"id":0,"name":"","addr":0,"addrs":null,"file":"","line":0,"Cond":"","continue":false,"traceReturn":false,"goroutine":false,"stacktrace":0,"LoadArgs":null,"LoadLocals":null,"hitCount":null,"totalHitCount":0}} error: "could not find file /go/bin/collector-linux/cmd/collector/app/server/http.go"
```

You can notice it's looking for code as per the binary path inside container.

Then I tried with Goland IDE and it worked well.

```
2020-10-13T13:25:58Z info layer=debugger created breakpoint: &api.Breakpoint{ID:21, Name:"", Addr:0xf43f41, Addrs:[]uint64{0xf43f41}, File:"github.com/jaegertracing/jaeger/cmd/collector/app/handler/http_handler.go", Line:59, FunctionName:"github.com/jaegertracing/jaeger/cmd/collector/app/handler.(*APIHandler).RegisterRoutes", Cond:"", Tracepoint:false, TraceReturn:false, Goroutine:false, Stacktrace:0, Variables:[]string(nil), LoadArgs:(*api.LoadConfig)(nil), LoadLocals:(*api.LoadConfig)(nil), HitCount:map[string]uint64{}, TotalHitCount:0x0}
```

These logs came after adding flag `--log-output=debugger,debuglineerr,gdbwire,lldbout,rpc` to the entry-point.

I have updated the branch with the above feedbacks related to CMD to ENTRYPOINT conversion and docker image name change. Thank you.

Reply:
I actually got it to work: 

```json
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Launch remote",
            "type": "go",
            "request": "attach",
            "mode": "remote",
            "remotePath": "",
            "port": 12345,
            "host": "127.0.0.1",
            "cwd": "${workspaceRoot}",
        }
    ]
}
```

I think it means you can go ahead :-)